File: leetcode/1202-smallest-string-with-swaps/Solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    private int[] parent;

    public String smallestStringWithSwaps(String s, List<List<Integer>> pairs) {
        int n = s.length();
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        for (List<Integer> pair : pairs) {
            union(pair.get(0), pair.get(1));
        }

        Map<Integer, List<Integer>> groups = new HashMap<>();
        for (int i = 0; i < n; i++) {
            groups.computeIfAbsent(find(i), k -> new ArrayList<>()).add(i);
        }

        char[] result = s.toCharArray();
        for (List<Integer> indices : groups.values()) {
            List<Integer> sortedIndices = new ArrayList<>(indices);
            Collections.sort(sortedIndices);

            List<Character> chars = new ArrayList<>();
            for (int i : sortedIndices) {
                chars.add(result[i]);
            }
            Collections.sort(chars);

            for (int k = 0; k < sortedIndices.size(); k++) {
                result[sortedIndices.get(k)] = chars.get(k);
            }
        }

        return new String(result);
    }

    private int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private void union(int a, int b) {
        int ra = find(a);
        int rb = find(b);
        if (ra != rb) {
            parent[ra] = rb;
        }
    }
}

File: leetcode/1202-smallest-string-with-swaps/solution.py
from collections import defaultdict
from typing import List


class Solution:
    def smallestStringWithSwaps(self, s: str, pairs: List[List[int]]) -> str:
        n = len(s)
        parent = list(range(n))

        def find(x: int) -> int:
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a: int, b: int) -> None:
            ra, rb = find(a), find(b)
            if ra != rb:
                parent[ra] = rb

        for a, b in pairs:
            union(a, b)

        groups = defaultdict(list)
        for i in range(n):
            groups[find(i)].append(i)

        result = list(s)
        for indices in groups.values():
            indices.sort()
            chars = sorted(result[i] for i in indices)
            for i, ch in zip(indices, chars):
                result[i] = ch

        return "".join(result)
